### PR TITLE
[wip] Normalize Ranges using RangeStart

### PR DESCRIPTION
**Pull request by btwj**

### CODE REVIEW

File: aeneas/src/core/Eval.v3 [hunk 1]
```diff
@@ -609,6 +609,43 @@ def evalOp(op: Operator, args: Arguments) -> Result {
 			if (val == null) return Int.ZERO;
 			return Int.box(ArrayRangeVal.!(val).length);
 		}
+		RangeStartPlusIndex => {
+			var start = ArrayRangeStart.!(args.vals[0]);
```
Comment (titzer) [inline aeneas/src/core/Eval.v3:613]: Should we make this work with `PointerRangeStart` as well?

File: test/bm.v3 [hunk 2]
```diff
@@ -0,0 +1,108 @@
+//@execute =1
```
Comment (titzer) [inline test/bm.v3:1]: This should probably live in `test/large`

File: aeneas/src/ssa/SsaGraph.v3 [hunk 3]
```diff
@@ -55,6 +55,13 @@ class SsaGraph {
 		}
 		return getConst(Long.TYPE, Long.box(v), Facts.longFacts(v));
 	}
+	def arrayRangeStartConst(v: int, arrayType: Type) -> SsaConst {
+		// TODO: Re-enable caching
```
Comment (titzer) [inline aeneas/src/ssa/SsaGraph.v3:59]: There's a general caching mechanism in `getConst`, you can just use that.

File: aeneas/src/main/CLOptions.v3 [hunk 4]
```diff
@@ -142,6 +142,8 @@ component CLOptions {
 		"Enable reachability analysis and normalization before running the program.");
 	def MA			= sharedOpt.newBoolOption("ma", true,
 		"Generate mixed arrays for arrays of tuples, rather than tuples of arrays.");
+	def NR			= sharedOpt.newBoolOption("nr", true,
```
Comment (titzer) [inline aeneas/src/main/CLOptions.v3:145]: We'll have to come up with better names for these flags, as this terseness was probably a bad idea in retrospect. Suggestions welcome.

File: aeneas/src/mach/MachProgram.v3 [hunk 5]
```diff
@@ -193,15 +194,22 @@ class MachProgram extends TargetProgram {
...
+			x: Record => {
+				if (V3.isVariant(x.rtype)) {
+					// check for variants represented as enums
+					if (prog.ir.isEnum(x.rtype)) return Int.box(V3.getVariantTag(x.rtype));
+				}
+				return addrOfRecord(x);
+			}
+			x: FuncVal => return addrOfMethod(x.memberRef.asMethod());
+			x: ArrayRangeStart => {
+				if (x.arrayType == null) return Int.box(x.start);
```
Comment (titzer) [inline aeneas/src/mach/MachProgram.v3:207]: How does it happen that the `arrayType` is null here?

File: aeneas/src/mach/MachLowering.v3 [hunk 6]
```diff
@@ -1761,8 +1799,11 @@ class MachLowering(mach: MachProgram, compiler: Compiler, config: MachLoweringCo
 				var left = base.inputs[0].dest, right = base.inputs[1].dest;
 				if (SsaConst.?(left)) {
 					// PtrAdd(PtrAdd(#left, right), #offset)
-					var addr = Addr.!(SsaConst.!(left).val);
-					left = context.graph.valConst(pd, addr.add(offset.unbox<int>()));
+					match (SsaConst.!(left).val) {
```
Comment (titzer) [inline aeneas/src/mach/MachLowering.v3:1802]: Uses of `SsaInstr.unbox` are getting suspicious these days. You might want to use `V3.unboxI32` (and also add a case for `ArrayRangeStart` there too.

File: aeneas/src/mach/MachLowering.v3 [hunk 7]
```diff
@@ -1452,28 +1467,51 @@ class MachLowering(mach: MachProgram, compiler: Compiler, config: MachLoweringCo
...
-		var arrayElemOffset = genArrayElemOffset(arrayRep.getElemElemOffset(0), 1, ni_offset);
+		var arrayElemOffset = if(oi.op.typeArgs[1] == V3Range.START_TYPE, start, genArrayElemOffset(arrayRep.getElemElemOffset(0), 1, start));
 		var machType = mach.machType(fieldType);
 		var loads = genNormTypedLoadsByteRegion(oi.source, machType, ptrAdd(arrayOffset, arrayElemOffset), 0);
 		mapN(oi, loads);
 	}
 	def genByteArraySetField(oi: SsaApplyOp, offset: int) {
 		var inputs = normRefs(oi.inputs);
-		var narr = inputs[0], ni_offset = inputs[1], nvalue = inputs[2];
+		var narr = inputs[0], start = inputs[1], nvalue = inputs[2];
```
Comment (titzer) [inline aeneas/src/mach/MachLowering.v3:1508]: Can you keep the `ni_` prefix here for now? I think I want to move all of this code to use `i_` or `ni_` prefixes for all variables that refer to graph nodes.

File: aeneas/src/mach/MachLowering.v3 [hunk 8]
```diff
@@ -1452,28 +1467,51 @@ class MachLowering(mach: MachProgram, compiler: Compiler, config: MachLoweringCo
 		oi.kill();
 		oi.remove();
 	}
+	def genNormRangeGetElem(oi: SsaApplyOp, elem: int) {
+		if (config.ObjectSystem) return void(normId(oi));
+		var inputs = normRefs(oi.inputs);
+		var narr = inputs[0], rangeStart = inputs[1], nindex = inputs[2], rangeType = oi.op.typeArgs[0];
```
Comment (titzer) [inline aeneas/src/mach/MachLowering.v3:1473]: Virgil desperately needs destructuring assignments :-)

**Review by titzer — COMMENTED** [review]
First round of comments.

File: test/bm.v3 (hunk 2 above)
Comment (btwj) [reply test/bm.v3:1]: Oops I was using this to track down a bug and accidentally committed it. Will come up with something smaller that reproduces the bug and add it into test/range!

File: aeneas/src/main/CLOptions.v3 (hunk 4 above)
Comment (btwj) [reply aeneas/src/main/CLOptions.v3:145]: Perhaps CLOptions should support long names with short aliases for options? Then we won't have to sacrifice clarity for terseness here.

File: aeneas/src/core/Opcode.v3 [hunk 9]
```diff
@@ -244,6 +253,13 @@ component Opcodes {
 		t[Opcode.RangeFromPlus.tag] = 		F;
 		t[Opcode.RangeGetLength.tag] = 		P;
 
+		t[Opcode.RangeStartPlusIndex.tag] =	P;
+		t[Opcode.RangeStartFromPointer.tag] =	P;
+		t[Opcode.NormRangeGetElem.tag] = 	F;
+		t[Opcode.NormRangeGetElemElem.tag] = 	F;
+		t[Opcode.NormRangeSetElem.tag] = 	NONE;
+		t[Opcode.NormRangeSetElemElem.tag] = 	NONE;
```
Comment (titzer) [inline aeneas/src/core/Opcode.v3:261]: I think `ForgeRange` and `PtrAddRangeStart` should be considered pure as well.

File: aeneas/src/mach/MachProgram.v3 (hunk 5 above)
Comment (btwj) [reply aeneas/src/mach/MachProgram.v3:207]: If a range has a default value, and we later use that default RangeStart (`null`) and do a `RangeStartPlusIndex` on it, the `RangeStartPlusIndex` won't know what `arrayType` to put in the constructed `ArrayRangeStart`.

File: aeneas/src/mach/MachLowering.v3 (hunk 6 above)
Comment (btwj) [reply aeneas/src/mach/MachLowering.v3:1802]: Should I replace all the `.unbox<int>()`'s in this section (including the old ones)?

Comment (titzer) [reply aeneas/src/mach/MachLowering.v3:1802]: Yep.

**Review by titzer — APPROVED** [review]
lgtm

**Pull request merged**